---
description: 刘小泽写于2021.5.25
---

# 2.3.2 算法 | 2021-scPhere——用地球仪来展示降维结果

> 常见的降维是把细胞落在2维空间平面图，这个工具比照地球仪进行细胞的降维

## 速览

* 作者：Jiarui Ding & Aviv Regev（通讯作者大家都熟知），来自Klarman Cell Observatory, Broad Institute of MIT and Harvard, Cambridge, MA, USA
* 链接：[https://www.nature.com/articles/s41467-021-22851-4](https://www.nature.com/articles/s41467-021-22851-4)
* 发表在：Nature Communications 
* 日期：2021-05-05
* 项目地址：[https://github.com/klarman-cell-observatory/scPhere](https://github.com/klarman-cell-observatory/scPhere)

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-023115.png)

## 摘要

* scRNA数据分析中，降维是非常重要的一步，用来解释各个细胞之间的亲疏远近，但目前的降维方案经常受到技术误差的影响，引起和真实生物差异的混淆
*   开发的scPhere（读作：sphere）可以在降维过程中区分出多层次复杂的批次效应，尤其对大型数据来说，结果不会把细胞们都堆在一起以至于看不出分化轨迹和细胞分群

    > 文章是这么定义这个工具的： scPhere, a scalable deep generative model to embed cells into low-dimensional hyperspherical or hyperbolic spaces to accurately represent scRNA-seq data
* 利用了9个大型数据集（包括病人、正在发育的动物）进行验证

## 首先看一下工具的模型

* 把scRNA的表达量和多个维度的批次信息读进来，批次信息包括了技术和生物两方面（比如不同的病人、疾病类型），在有批次效应的情况下，学习细胞潜在的数据结构
* 学习得到的模型可以被用来：（1）找某个生物因素（如疾病）对表达量的影响；（2）得到一个不受批次效应影响的参考数据，然后把新的数据集去和它比较；（3）看看细胞的空间分布

![Fig. 1](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-023928.png)

## 用小型和大型数据集比较不同的降维方法

采用的数据集是：

* “**small**” datasets were: (1) a blood cell dataset with only 10 erythroid cell profiles and 2293 CD14+ monocytes; (2) 3314 human lung cells, (3) 1378 mouse white adipose tissue stromal cells, and (4) 1755 human splenic nature killer cells spanning four subtypes
* “**large**” datasets were: (1) **35,699** retinal ganglion cells in 45 cell subsets; and (2) **599,926** cells spanning 102 subsets across 59 human tissues in the Human Cell Landscape

比较对象是：t-SNE, UMAP, and PHATE

图中上面8张是小数据集，下面8张是大型数据集

* 对小数据集来讲，t-SNE, UMAP, and PHATE的降维结果都还不错，都基本没有批次效应
* 当切换到大型数据集，scPhere的优势就体现出来了：保证了数据全局的层级结构，能将同属一个大群的不同亚群聚在一起
* 当数据量增大，t-SNE的图变得越来越”臃肿“，在2D图中，即使是非常不同的细胞类型，也会被”挤“得非常近（图k）；而UMAP中多个cluster出现混乱的情况（图I）

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-030909.png)

*   当然，PHATE对大型数据集的处理结果不理想

    ![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-032926.png)

## 测试scPhere的批次效应处理性能

> 大部分的批次效应处理工具只能处理单个的批次信息，而scPhere的特性是能结合多个批次信息进行学习

测试数据集：**301,749** cells we previously profiled in a complex experimental design from the colon mucosa of **18 patients** with ulcerative colitis (UC 溃疡性结肠炎), a major type of inflammatory bowel diseases (IBD 发炎性肠症), and **12 healthy individuals**

其中除了病人这个批次信息，还有：

* individuals were either healthy or with UC
* cells were collected separately from the epithelial 
* lamina propria fractions of each biopsy
* two replicate biopsies for each healthy individual 
* samples were collected at two time periods, separated by over a year

比较工具：Harmony, LIGER, and Seurat3 CCA（因为后两个只能处理一个批次信息，就选取了不同的人作为批次）

得到结论：**scPhere’s batch correction** on this complex dataset (30 patients with disease and location factors) performed **better than Harmony, Seurat3 CCA, and LIGER** based on classification accuracies of cell types for stromal, epithelial, and immune cells

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-043453.png)

图n和o都是scPhere的两个展示方式，分别是Embedding和Equal Earth map projection，对300,000个stromal, epithelial, and immune cells进行降维。其中还加入了是否患病、疾病类型、患病位点作为批次信息。

图p、q是利用Harmony的批次处理结果，分别采用t-SNE和UMAP的降维，结果就不理想。比如WNT2B+ fibroblasts, RSPO3+ fibroblasts 以及inflammatory fibroblasts就无法区分，而plasma cells就被莫名其妙地拆分，其他不同谱系的细胞竟然也离得很近

## 用scPhere帮助轨迹推断

when we quantify time continuity, by comparing the k-nearest neighbor time point classification accuracies, accuracies from **scPhere** (in 2D) were **higher** than those from t-SNE, UMAP, and PHATE

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-052222.png)

## 总结

几大特性：

* **Accounting for multilevel complex batch effects**: ScPhere’s ability to handle complex batch factors is an advantage over previous methods for batch correction (e.g., SAUCIE, scVI, LIGER, Seurat3 CCA, fastMNN, Scanorama, and Conos), which handle only one batch vector.
* Especially useful for **analyzing large scRNA-seq datasets**: does not suffer from “cell-crowding” even with large numbers of input cells; better preserves hierarchical, global structures;  forms a reference to annotate new profiled cells from future studies (这一点对大型项目非常有帮助，比如健康群体的Human Cell Atlas项目，疾病群体的Human Tumor Atlas Network项目，都需要构建一个reference map)
* modifying it to spatially map cells

未来扩展：

* include semi-supervised learning to **annotate** cell types
* imputing **missing counts** in scRNA-seq data and removing ambient RNA **contamination**
* integrative analysis of **multimodal** data (e.g. spatial transcriptomics, single-cell ATAC-seq)
* learn **discrete hierarchical trees** for betterd interpreting developmental trajectories
* model perturbation data

最后看一下这种降维结果的动画：

![](https://jieandze1314-1255603621.cos.ap-guangzhou.myqcloud.com/blog/2021-05-25-053726.gif)
